packages/backend-lalr/data/HappyTemplate.hs

```diff
@@ -85,145 +60,153 @@ happyParse start_state = happyNewToken start_state notHappyAtAll notHappyAtAll
-         off_i  = PLUS(off, i)
-         check  = if GTE(off_i,(ILIT(0) :: FAST_INT))
-                  then EQ(indexShortOffAddr happyCheck off_i, i)
-                  else Prelude.False
-         action
-          | check     = indexShortOffAddr happyTable off_i
-          | Prelude.otherwise = indexShortOffAddr happyDefActions st
-
-#ifdef HAPPY_GHC
-indexShortOffAddr (HappyA# arr) off =
-        Happy_GHC_Exts.narrow16Int# i
+happyDoAction i tk st =
+  DEBUG_TRACE("state: " ++ show (Happy_GHC_Exts.I# st) ++
+              ",\ttoken: " ++ show (Happy_GHC_Exts.I# i) ++
+              ",\taction: ")
+  case happyDecodeAction (happyNextAction i st) of
+    HappyFail             -> DEBUG_TRACE("failing.\n")
+                             happyFail (happyExpListPerState (Happy_GHC_Exts.I# st)) i tk st
+    HappyAccept           -> DEBUG_TRACE("accept.\n")
+                             happyAccept i tk st
+    HappyReduce rule      -> DEBUG_TRACE("reduce (rule " ++ show (Happy_GHC_Exts.I# rule) ++ ")")
+                             (happyReduceArr Happy_Data_Array.! (Happy_GHC_Exts.I# rule)) i tk st
+    HappyShift  new_state -> DEBUG_TRACE("shift, enter state " ++ show (Happy_GHC_Exts.I# new_state) ++ "\n")
+                             happyShift new_state i tk st
+
+{-# INLINE happyNextAction #-}
+happyNextAction i st = case happyIndexActionTable i st of
+  Just (Happy_GHC_Exts.I# act) -> act
+  Nothing                      -> indexOffAddr happyDefActions st
+
+{-# INLINE happyIndexActionTable #-}
+happyIndexActionTable i st
+  | GTE(off, 0#), EQ(indexOffAddr happyCheck off, i)
+  = Prelude.Just (Happy_GHC_Exts.I# (indexOffAddr happyTable off))
+  | otherwise
+  = Prelude.Nothing
   where
-        i = Happy_GHC_Exts.word2Int# (Happy_GHC_Exts.or# (Happy_GHC_Exts.uncheckedShiftL# high 8#) low)
-        high = Happy_GHC_Exts.int2Word# (Happy_GHC_Exts.ord# (Happy_GHC_Exts.indexCharOffAddr# arr (off' Happy_GHC_Exts.+# 1#)))
-        low  = Happy_GHC_Exts.int2Word# (Happy_GHC_Exts.ord# (Happy_GHC_Exts.indexCharOffAddr# arr off'))
-        off' = off Happy_GHC_Exts.*# 2#
-#else
-indexShortOffAddr arr off = arr Happy_Data_Array.! off
-#endif
+    off = PLUS(indexOffAddr happyActOffsets st, i)
+
+data HappyAction
+  = HappyFail
+  | HappyAccept
+  | HappyReduce Happy_Int -- rule number
+  | HappyShift Happy_Int  -- new state
+
+{-# INLINE happyDecodeAction #-}
+happyDecodeAction :: Happy_Int -> HappyAction
+happyDecodeAction  0#                        = HappyFail
+happyDecodeAction -1#                        = HappyAccept
+happyDecodeAction action | LT(action, 0#)    = HappyReduce NEGATE(PLUS(action, 1#))
+                         | otherwise         = HappyShift MINUS(action, 1#)
+
+{-# INLINE happyIndexGotoTable #-}
+happyIndexGotoTable nt st = indexOffAddr happyTable off
+  where
+    off = PLUS(indexOffAddr happyGotoOffsets st, nt)
+
+
+indexOffAddr (HappyA# arr) off =
+  Happy_GHC_Exts.int32ToInt# (Happy_GHC_Exts.indexInt32OffAddr# arr off)
```

Comment:
> Should we measure?

Yes. In my experience endianness-specific optimizations of this nature can have significant performance impact.

Reply:
We just recently measured this in https://github.com/haskell/tar/pull/95:
* yes, it matter a lot,
* use `byteSwap32` for BE.

Reply:
Thanks for the tip, @Bodigrim. I opened a PR on `alex` to check out whether we get the intended semantics: https://github.com/haskell/alex/pull/260

It's hard to measure for me, though, because I do not have a big endian device with GHC.

The best I could do would be to swap the order in `ProduceCode.lhs` and measure `byteSwap32#` on x86\_64. I'm not sure if that's a useful thing to do.

Provided we do the `byteSwap32#` thing, would you be satisfied without measurements? I think we'll be faster on x86\_64 and aarch64 anyway, because we no longer do 2 separate accesses and bit shifts.

Reply:
`byteSwap32#` is a single CPU instruction and it's extremely hard to get access to real big-endian hardware, so it's fine to proceed without measurements. Setting up a big-endian CI job would be very useful though.

Reply:
I added the `byteSwapX#` code to `alex` and added a CI job there to test the cod works as intended.

Unfortunately, setting up an emulated CI job for `happy` is a tad too painful for my tastes, because we cannot use `cabal` to build the project and `happy` is split in multiple different projects.

Since I made sure to literally just copy and paste the code from `alex`, I'm fairly certain that we can merge as is.